### ISSUE TITLE
[CI] Rolling upgrade tests failing on 7.x due to node throwing uncaught AssertionError

### ISSUE THREAD

**Issue by mark-vieira** [elastic/elasticsearch#66399]
RollupDateHistoUpgradeIT.testDateHistoIntervalUpgrade seems to be failing in our rolling upgrade tests as of Dec 14th.

We're getting a number of builds failing here and in them I see this asserition error in one of the nodes:

```
» ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v7.10.2-2] fatal error in thread [elasticsearch[v7.10.2-2][generic][T#1]], exiting
»  java.lang.AssertionError: null
»  	at org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:55) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:145) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$DiffableValueReader.read(DiffableUtils.java:688) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$DiffableValueReader.read(DiffableUtils.java:677) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$MapDiff.<init>(DiffableUtils.java:408) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$JdkMapDiff.<init>(DiffableUtils.java:178) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils.readJdkMapDiff(DiffableUtils.java:167) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStreamMetadata$DataStreamMetadataDiff.<init>(DataStreamMetadata.java:169) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStreamMetadata.readDiffFrom(DataStreamMetadata.java:82) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:46) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.NamedDiffableValueSerializer.readDiff(NamedDiffableValueSerializer.java:56) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.NamedDiffableValueSerializer.readDiff(NamedDiffableValueSerializer.java:30) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$MapDiff.<init>(DiffableUtils.java:401) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$ImmutableOpenMapDiff.<init>(DiffableUtils.java:234) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils.readImmutableOpenMapDiff(DiffableUtils.java:127) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.Metadata$MetadataDiff.<init>(Metadata.java:951) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.Metadata.readDiffFrom(Metadata.java:887) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.ClusterState$ClusterStateDiff.<init>(ClusterState.java:787) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.ClusterState.readDiffFrom(ClusterState.java:697) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.coordination.PublicationTransportHandler.handleIncomingPublishRequest(PublicationTransportHandler.java:197) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.coordination.PublicationTransportHandler.lambda$new$0(PublicationTransportHandler.java:102) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:257) ~[?:?]
»  	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:224) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.lambda$messageReceived$0(SecurityServerTransportInterceptor.java:306) ~[?:?]» ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v7.10.2-2] fatal error in thread [elasticsearch[v7.10.2-2][generic][T#1]], exiting
»  java.lang.AssertionError: null
»  	at org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:55) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:145) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$DiffableValueReader.read(DiffableUtils.java:688) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$DiffableValueReader.read(DiffableUtils.java:677) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$MapDiff.<init>(DiffableUtils.java:408) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$JdkMapDiff.<init>(DiffableUtils.java:178) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils.readJdkMapDiff(DiffableUtils.java:167) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStreamMetadata$DataStreamMetadataDiff.<init>(DataStreamMetadata.java:169) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.DataStreamMetadata.readDiffFrom(DataStreamMetadata.java:82) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:46) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.NamedDiffableValueSerializer.readDiff(NamedDiffableValueSerializer.java:56) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.NamedDiffableValueSerializer.readDiff(NamedDiffableValueSerializer.java:30) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$MapDiff.<init>(DiffableUtils.java:401) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils$ImmutableOpenMapDiff.<init>(DiffableUtils.java:234) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.DiffableUtils.readImmutableOpenMapDiff(DiffableUtils.java:127) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.Metadata$MetadataDiff.<init>(Metadata.java:951) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.metadata.Metadata.readDiffFrom(Metadata.java:887) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.ClusterState$ClusterStateDiff.<init>(ClusterState.java:787) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.ClusterState.readDiffFrom(ClusterState.java:697) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.coordination.PublicationTransportHandler.handleIncomingPublishRequest(PublicationTransportHandler.java:197) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.cluster.coordination.PublicationTransportHandler.lambda$new$0(PublicationTransportHandler.java:102) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:257) ~[?:?]
»  	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:224) ~[elasticsearch-7.10.2-SNAPSHOT.jar:7.10.2-SNAPSHOT]
»  	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.lambda$messageReceived$0(SecurityServerTransportInterceptor.java:306) ~[?:?]
```

https://gradle-enterprise.elastic.co/s/oggowvflnvbxy

Here's a history of these failures;
https://gradle-enterprise.elastic.co/scans/failures?failures.failureClassification=all_failures&failures.failureMessage=Execution%20failed%20for%20task%20%27:x-pack:qa:rolling-upgrade:v*%23twoThirdsUpgradedTest%27.%0A%3E%20There%20were%20failing%20tests.%20See%20the%20report%20at:%20file:///*/*/*/*/build/reports/tests/*/index.html&search.relativeStartTime=P7D&search.tags=CI&search.tags=not:pull-request&search.tags=7.x&search.timeZoneId=America/Los_Angeles

**Comment by mark-vieira** [comment]
@danhermann Looking through commit history you've done some work in the area of the stacktrace in 7.x recently. Since this is BWC stuff no idea if it's at all related but perhaps you have some idea what's going on here?

**Comment by danhermann** [comment]
This looks like https://github.com/elastic/elasticsearch/issues/66370. BWC tests were muted in master and 7.x this morning until the (under review) fix for that can be merged. Does that address the build issues for now?

**Comment by mark-vieira** [comment]
Yep, I believe so. Thanks!

**Comment by mark-vieira** [comment]
@danhermann is this related or something else:
https://gradle-enterprise.elastic.co/s/od4dieroi65mg/tests/:x-pack:qa:full-cluster-restart:v7.11.0%23oldClusterTest/org.elasticsearch.xpack.restart.FullClusterRestartIT/testDataStreams#1

**Comment by danhermann** [comment]
It's related but not exactly the same. I will mute it until it can be fixed.

**Comment by mark-vieira** [comment]
> It's related but not exactly the same. I will mute it until it can be fixed.

Since BWC tests are disabled should be no need to mute. I just saw that in an out of date branch. Unless #66370 won't resolve that specific issue and it'll need to be fixed separately.

**Comment by danhermann** [comment]
> > It's related but not exactly the same. I will mute it until it can be fixed.
> 
> Since BWC tests are disabled should be no need to mute. I just saw that in an out of date branch. Unless #66370 won't resolve that specific issue and it'll need to be fixed separately.

Oh, I see. I expected https://github.com/elastic/elasticsearch/pull/66316 to fix that one and didn't see that it was an out-of-date branch.

**Comment by danhermann** [comment]
This should be resolved by #66400 and its backports.
